### PR TITLE
[FEAT] Predict insample

### PR DESCRIPTION
**Pull request by cchallu**
- Added `predict_insample` method to `core`. Will return historic forecasts for the complete training set (defined first with `fit` or `cross_validation`). The function will automatically trim the test set in the dataset.
- Added `SAMPLING_TYPE` class attribute to each model.
- Added docstring to base classes.
- Added `trim_dataset` static method to `TimeSeriesDataset`.

### CODE REVIEW

File: neuralforecast/common/_base_windows.py [hunk 1]
```diff
@@ -234,9 +245,17 @@ def _create_windows(self, batch, step):
 
         elif step in ["predict", "val"]:
             if step == "predict":
+                initial_input = temporal.shape[-1] - self.test_size
+                if (
+                    initial_input <= self.input_size
+                ):  # There is not enough data to predict first timestamp
+                    padder_left = nn.ConstantPad1d(
```
Comment (kdgutier) [inline neuralforecast/common/_base_windows.py:252]: Ibid 
Comment on missing work to add encoded static features into the initial condition of the encoders.

File: neuralforecast/core.py [hunk 2]
```diff
@@ -72,7 +72,25 @@ def _cv_dates(last_dates, freq, h, test_size, step_size=1):
         dates = dates.reset_index(drop=True)
     return dates
 
-# %% ../nbs/core.ipynb 9
+# %% ../nbs/core.ipynb 6
+def _insample_dates(uids, last_dates, freq, h, len_series, step_size=1):
+    if (len(np.unique(last_dates)) == 1) and (len(np.unique(len_series)) == 1):
```
Comment (kdgutier) [inline neuralforecast/core.py:77]: Can we add a brief description of what `_insample_dates` method does?

File: neuralforecast/core.py [hunk 3]
```diff
@@ -422,6 +451,99 @@ def cross_validation(
         fcsts_df = fcsts_df.merge(df, how="left", on=["unique_id", "ds"])
         return fcsts_df
 
+    def predict_insample(self, step_size: int = 1):
```
Comment (kdgutier) [inline neuralforecast/core.py:454]: Can we rename the function inspired by StatsForecast method?
We can also match the documentation with StatsForecast's.

https://nixtla.github.io/statsforecast/core.html#statsforecast.forecast_fitted_values

File: neuralforecast/models/autoformer.py [hunk 4]
```diff
@@ -475,6 +475,9 @@ class Autoformer(BaseWindows):
         - [Wu, Haixu, Jiehui Xu, Jianmin Wang, and Mingsheng Long. "Autoformer: Decomposition transformers with auto-correlation for long-term series forecasting"](https://proceedings.neurips.cc/paper/2021/hash/bcc0d400288793e8bdcd7c19a8ac0c2b-Abstract.html)<br>
     """
 
+    # Class attributes
+    SAMPLING_TYPE = "windows"
```
Comment (kdgutier) [inline neuralforecast/models/autoformer.py:479]: Why `SAMPLING_TYPE` in caps.

File: neuralforecast/tsdataset.py [hunk 5]
```diff
@@ -199,6 +199,53 @@ def update_dataset(dataset, future_df):
 
         return updated_dataset
 
+    @staticmethod
+    def trim_dataset(dataset, left_trim: int = 0, right_trim: int = 0):
+        """Trim temporal information from a dataset."""
```
Comment (kdgutier) [inline neuralforecast/tsdataset.py:204]: """
Trim temporal information from a dataset.
Keeping window indexes [t+left:t-right] for all series.
"""

File: neuralforecast/core.py [hunk 6]
```diff
@@ -422,6 +451,99 @@ def cross_validation(
...
+                warnings.warn(
+                    f"Predict insample might not provide accurate predictions for \
+                       recurrent model {repr(model)} yet due to scaling."
+                )
+                print(
+                    f"WARNING: Predict insample might not provide accurate predictions for \
+                      recurrent model {repr(model)} yet due to scaling."
+                )
+
+        cols = []
```
Comment (kdgutier) [inline neuralforecast/core.py:486]: Add comment explaining that we are covering repeated model apparitions with a suffix.

File: neuralforecast/core.py [hunk 7]
```diff
@@ -422,6 +451,99 @@ def cross_validation(
...
+        count_names = {"model": 0}
+        for model in self.models:
+            model_name = repr(model)
+            count_names[model_name] = count_names.get(model_name, -1) + 1
+            if count_names[model_name] > 0:
+                model_name += str(count_names[model_name])
+            cols += [model_name + n for n in model.loss.output_names]
+
+        # Remove test set from dataset and last dates
+        test_size = self.models[0].get_test_size()
```
Comment (kdgutier) [inline neuralforecast/core.py:496]: Do we need to protect all models having the same `test_size`?
I think this can have weird unintended interactions.

File: neuralforecast/core.py [hunk 8]
```diff
@@ -422,6 +451,99 @@ def cross_validation(
...
+        if test_size > 0:
+            dataset = TimeSeriesDataset.trim_dataset(
```
Comment (kdgutier) [inline neuralforecast/core.py:498]: Can we rename `dataset` to `trimmed_dataset`?
That way is easier to follow the intuition of the code.

**Review by kdgutier — CHANGES_REQUESTED** [review]

File: neuralforecast/core.py (hunk 7 above)
Comment (cchallu) [reply neuralforecast/core.py:496]: yes, test_size is fixed for all models by the `cross_validation` method

File: neuralforecast/common/_base_auto.py [hunk 9]
```diff
@@ -152,6 +152,9 @@ def __init__(
         self.valid_loss = self.config.get("valid_loss", self.loss)
         self.alias = alias
 
+        # Base Class attributes
+        self.SAMPLING_TYPE = cls_model.SAMPLING_TYPE
```
[the remark this answers is not in the document]
Comment (cchallu) [reply neuralforecast/common/_base_auto.py:156]: naming convention of static attributes

File: neuralforecast/core.py [hunk 10]
```diff
@@ -422,6 +451,99 @@ def cross_validation(
...
+
+        for model in self.models:
+            if model.SAMPLING_TYPE == "recurrent":
+                warnings.warn(
+                    f"Predict insample might not provide accurate predictions for \
+                       recurrent model {repr(model)} yet due to scaling."
+                )
+                print(
+                    f"WARNING: Predict insample might not provide accurate predictions for \
+                      recurrent model {repr(model)} yet due to scaling."
```
Comment (FedericoGarza) [inline neuralforecast/core.py]: Do you know if these lines are repeated? I prefer using `warning.warn` but it is just a preference. Not blocking.

File: neuralforecast/core.py [hunk 11]
```diff
@@ -463,6 +585,9 @@ def predict_rolled(
             DataFrame with insample `models` columns for point predictions and probabilistic
             predictions for all fitted `models`.
         """
+        print(
+            "WARNING: this method will be deprecated. Use `cross_validation` or `predict_insample` instead."
+        )
```
Comment (FedericoGarza) [inline neuralforecast/core.py:594]: I prefer using `warning.warn` instead of `print` but it is just a preference. Not blocking.

File: neuralforecast/common/_base_windows.py (hunk 1 above)
Comment (FedericoGarza) [reply neuralforecast/common/_base_windows.py:252]: I prefer to add pending work on issues instead of directly in the code. But just a preference.

**Review by FedericoGarza — APPROVED** [review]
Nice feature! I left a couple of non-blocking comments. LGTM :)

**Review by kdgutier — APPROVED** [review]

**Comment by candalfigomoro** [comment]
@cchallu @FedericoGarza @kdgutier 
Can this fix https://github.com/Nixtla/hierarchicalforecast/issues/179 ?
